### PR TITLE
In-Memory Registry

### PR DESCRIPTION
**Pull request by Rostifar**
# In-Memory Registry 
Implements an in-memory registry, which stores all FeastObjects in memory rather than persisting to an external sink.

### CODE REVIEW

File: sdk/python/feast/repo_operations.py [hunk 1]
```diff
@@ -114,6 +115,7 @@ def parse_repo(repo_root: Path) -> RepoContents:
         request_feature_views=[],
     )
 
+    res.entities.append(DUMMY_ENTITY)
```
Comment (maksstach) [inline sdk/python/feast/repo_operations.py:118]: @Rostifar why is this needed?

Comment (Rostifar) [reply sdk/python/feast/repo_operations.py:118]: Yeah, this is an entity implemented for FeatureViews that don't have entities: checkout this https://github.com/Affirm/feast/blob/de3d7b8d36c8c5c08e372c3dbb078d17e8870274/sdk/python/feast/repo_operations.py#L201

File: sdk/python/feast/infra/registry/memory.py [hunk 2]
```diff
@@ -0,0 +1,758 @@
...
+    if len(s) != 2:
+        raise ValueError(f"Invalid projected key {projected_key}. Key must follow the format `project:name`.")
+    return s[0], s[1]
+
+
+def list_registry_dict(project: str, registry: Dict[str, FeastResource]) -> List[FeastResource]:
+    # returns all values in registry that belong to `project`
+    return [
+        v for k, v in registry.items() if invert_projected_key(k)[0] == project
+    ]
```
Comment (maksstach) [inline sdk/python/feast/infra/registry/memory.py:75]: Why did you decide to keep the registry flat and have a dict for every FeastResource + projected key instead of having a dict of dicts that we would call like `self.entities["project_1"]["entity_x"]`?

File: sdk/python/feast/infra/registry/memory.py [hunk 3]
```diff
@@ -0,0 +1,758 @@
...
+        # not supported for BaseFeatureView in-memory objects
+        pass
+
+    def get_user_metadata(self, project: str, feature_view: BaseFeatureView) -> Optional[bytes]:
+        # not supported for BaseFeatureView in-memory objects
+        pass
+
+    def proto(self) -> RegistryProto:
+        if self.cached_proto:
+            return self.cached_proto
```
Comment (maksstach) [inline sdk/python/feast/infra/registry/memory.py:716]: nice

File: sdk/python/feast/infra/registry/memory.py (hunk 2 above)
Comment (Rostifar) [reply sdk/python/feast/infra/registry/memory.py:75]: Yeah, we can nest this. I don't have a strong preference, I think.

File: sdk/python/feast/errors.py [hunk 4]
```diff
@@ -111,13 +123,25 @@ def __init__(self, name: str, project: str):
...
     def __init__(self, name: str, project: str):
         super().__init__(
             f"Validation reference {name} does not exist in project {project}"
         )
 
 
+class DuplicateValidationReference(Exception):
+    def __init__(self, name: str, project) -> None:
+        super(DuplicateValidationReference, self).__init__(
+            f"Duplication validation reference {name} for project {project}."
```
Comment (dominicbecker) [inline sdk/python/feast/errors.py:141]: nit
```suggestion
            f"Duplicate validation reference {name} for project {project}."
```

File: sdk/python/feast/repo_config.py [hunk 5]
```diff
@@ -165,6 +165,9 @@ class RepoConfig(FeastBaseModel):
     coerce_tz_aware: Optional[bool] = True
     """ If True, coerces entity_df timestamp columns to be timezone aware (to UTC by default). """
 
+    ignore_infra_changes: bool = False
+    """ Signals whether to apply infrastructure changes during Registry creation """
```
Comment (dominicbecker) [inline sdk/python/feast/repo_config.py:169]: when would one want to use this flag? mind adding brief details here?

**Review by dominicbecker — APPROVED** [review]
overall looks great. 

are there any automated tests?

File: sdk/python/feast/repo_config.py (hunk 5 above)
Comment (Rostifar) [reply sdk/python/feast/repo_config.py:169]: @dominicbecker 
We need to support the following: infrastructure changes should only be applied once during a deploy. With every `FeatureStore` instance, the registry is created from scratch. Without extra information, there's no indication when or where we should apply infrastructure changes.

Comment (dominicbecker) [reply sdk/python/feast/repo_config.py]: do you mind updating the docstring with some more info?

e.g. include that it only applies for `registry_type: memory`

Comment (Rostifar) [reply sdk/python/feast/repo_config.py]: Updated
